File: frontend/src/app/ticketService.ts
import { apiRequest } from './api';

export interface BackendTicket {
  id: string;
  vehiclePlate: string;
  ownerName: string;
  slot: string;
  entryTime: string;
  exitTime?: string;
  amount?: number;
  status: 'ONGOING' | 'FINISHED';
  vehicleType: string;
  paymentMethod?: string;
  ratePerHour?: number;
  durationMins?: number;
  zone?: string;
}

export const ticketService = {
  fetchTickets: () =>
    apiRequest<BackendTicket[]>('/tickets'),

  fetchTicket: (id: string) =>
    apiRequest<BackendTicket>(`/tickets/${id}`),

  createTicket: (ticket: BackendTicket) =>
    apiRequest<BackendTicket>('/tickets', {
      method: 'POST',
      body: JSON.stringify(ticket),
    }),

  updateTicket: (id: string, ticket: BackendTicket) =>
    apiRequest<BackendTicket>(`/tickets/${id}`, {
      method: 'PUT',
      body: JSON.stringify(ticket),
    }),

  deleteTicket: (id: string) =>
    apiRequest<void>(`/tickets/${id}`, {
      method: 'DELETE',
    }),

  checkoutTicket: (id: string, paymentMethod: string) =>
    apiRequest<BackendTicket>(`/tickets/${id}/checkout`, {
      method: 'POST',
      params: { paymentMethod },
    }),
};

File: backend/src/main/java/com/parkpulse/ticket/service/TicketService.java
package com.parkpulse.ticket.service;

import com.parkpulse.ticket.dto.TicketDTO;
import java.util.List;
import java.util.Optional;

/**
 * Abstraction for Ticket business logic.
 */
public interface TicketService {
    List<TicketDTO> getAllTickets(
            String status,
            String vehicleType,
            String paymentMethod,
            String slot,
            String dateFrom,
            String dateTo,
            Double amountMin,
            Double amountMax);

    Optional<TicketDTO> getTicketById(String id);
    TicketDTO createTicket(TicketDTO ticketDTO);
    TicketDTO updateTicket(String id, TicketDTO ticketDTO);
    void deleteTicket(String id);
    TicketDTO checkout(String id, String paymentMethod);
    TicketDTO checkout(String id, String paymentMethod, Double ratePerHour);
    TicketDTO createFromParkingSlot(String vehiclePlate, String ownerName, String slot, String vehicleType,
                                    java.time.LocalDateTime entryTime, String paymentMethod, double ratePerHour);
    TicketDTO createOngoingFromParking(String vehiclePlate, String ownerName, String slot, String vehicleType,
                                       java.time.LocalDateTime entryTime, double ratePerHour);
    java.util.Optional<TicketDTO> findOngoingTicketBySlot(String slotId);
}
